### ISSUE TITLE
Apple with M1 Pro outputs 0 stats when training with MPS

### ISSUE THREAD

**Issue by andreivicol** [ultralytics/ultralytics#9113]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

Training my custom dataset on M1 Pro with MPS as parameter reports 0 for all statistical metrics after each epoch. Training on CPU works as expected, giving real values that improve with each epoch. The code I use is:

When training on GPU, the console outputs the following:
<img width="1048" alt="appleMPSfail" src="https://github.com/ultralytics/ultralytics/assets/63958977/3bc893a7-e320-4495-860e-b76d4c59e835">

When using the CPU, it works as expected:
<img width="1047" alt="appleCPUok" src="https://github.com/ultralytics/ultralytics/assets/63958977/3dc90ccf-7e2b-45ed-8209-53380936e95a">

I can see that the training time on GPU Is shorter, thus I imagine it is actually running on the GPU as it should. I cannot see why the CPU would run Ok but the GPU wouldn't. Environment is setup with anaconda and I also have PyTorch for apple silicon.

### Environment

OS                  macOS-14.4-arm64-arm-64bit
Environment         Darwin
Python              3.12.2
Install             pip
RAM                 16.00 GB
CPU                 Apple M1 Pro
CUDA                None

matplotlib          ✅ 3.8.0>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.11.4>=1.4.1
torch               ✅ 2.2.1>=1.8.0
torchvision         ✅ 0.17.1>=0.9.0
tqdm                ✅ 4.65.0>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.1>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0

### Minimal Reproducible Example

from ultralytics import YOLO
from datetime import datetime, timedelta

model = YOLO('yolov8n.pt')

startTime = datetime.now()
results = model.train(data='/data.yaml', epochs=15,
                      imgsz=[480, 352], device="mps") # have also tried with 'mps:0' to no avail
endTime = datetime.now()
delta = endTime - startTime
print('\n\n\n')
print(f'Model training took {delta}')
print(f'Model training took {delta.seconds}')

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@andreivicol hey there! 😊 It looks like you're facing an issue with training on Apple M1 Pro using Metal Performance Shaders (MPS). This is quite peculiar, especially since CPU training works fine but MPS shows zeros across statistical metrics.

Given everything checks out with your environment setup and dependencies, let's try a couple of things:

1. Ensure PyTorch for apple silicon is up-to-date and specifically optimized for MPS. Sometimes, a version mismatch or update can address these issues.
2. Explicitly define your `device` argument as just `"mps"`, without the index. It sounds like you've already attempted this, but double-checking never hurts.
3. Inspect the MPS library for any known issues or updates regarding statistical metric reporting during training.

Here's a slightly adjusted snippet from your example to reiterate the points above:

```python
from ultralytics import YOLO
from datetime import datetime

model = YOLO('yolov8n.pt')

start_time = datetime.now()
results = model.train(data='/data.yaml', epochs=15, imgsz=[480, 352], device="mps")
end_time = datetime.now()
delta = end_time - start_time

print(f'\nModel training took {delta}')
```

If these steps still do not solve the issue, it might be helpful to create a virtual environment specifically for testing this setup to ensure no global package discrepancies are affecting the MPS performance.

Also, please keep an eye on our [GitHub repository](https://github.com/ultralytics/ultralytics) for updates or similar reports from users with Apple silicon. It's possible someone else has encountered this and found a workaround or fix. 👀

Let us know how it goes!

**Comment by mysakbm** [comment]
I have the same issue. I'm using M1 Pro. When I trained ono coco dataset, it is ok. However, when I tried custom dataset with 2500 images, training on CPU is ok, with "MPS" displays nans...

`import os
from ultralytics import YOLO
from ultralytics import settings

os.chdir("image_recognition/object_detection/yolo")

# Load a model
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)

# Train the model with 2 GPUs
results = model.train(
    data='../datasets/signatures_dataset/SignverOD/signverOD.yaml',
    epochs=1,
    device='mps'
)
`

Running this code. With this error:
![image](https://github.com/ultralytics/ultralytics/assets/38407239/b82e878a-a4f8-4fae-83be-83ea5e398468)

**Comment by glenn-jocher** [comment]
@mysakbm hey there! 👋 It seems like you've encountered the infamous NaN issue when training on MPS with a custom dataset. Since training on the COCO dataset works fine, but your custom dataset leads to NaN values, this typically hints at issues such as label formatting or data preprocessing discrepancies specific to your custom dataset.

Here are a couple of things you could check:

- **Label Format:** Make sure your labels are correctly formatted according to YOLOv8 requirements. Label issues are a common cause of NaNs during training.
- **Normalization:** Verify that your custom dataset images and annotations are properly normalized and match the preprocessing expectations for YOLOv8.
- **Batch Size and Learning Rate:** Sometimes, starting with a smaller learning rate or adjusting the batch size can help stabilize training and mitigate NaN issues.

As for MPS-specific suggestions:

- Double-check if all tensors are moved to the MPS device as expected. MPS can be a bit finicky about where tensors reside.
- Look for any MPS-specific limitations or requirements in the Ultralytics documentation or PyTorch release notes.

Here’s a small tip for debugging, try isolating a small subset of your data (maybe just a few images) and see if training proceeds without NaNs. This can often point towards issues in the broader dataset.

Since you're using an M1 Pro, ensure your environment is up-to-date, and there are no conflicts between TensorFlow and PyTorch libraries, as your snippet seems to import both.

Remember, the community and Ultralytics team are always here to help. If the issue persists, feel free to share more details or reach out on the discussion boards! 💪

**Comment by mysakbm** [comment]
> @mysakbm hey there! 👋 It seems like you've encountered the infamous NaN issue when training on MPS with a custom dataset. Since training on the COCO dataset works fine, but your custom dataset leads to NaN values, this typically hints at issues such as label formatting or data preprocessing discrepancies specific to your custom dataset.
> 
> Here are a couple of things you could check:
> 
> * **Label Format:** Make sure your labels are correctly formatted according to YOLOv8 requirements. Label issues are a common cause of NaNs during training.
> * **Normalization:** Verify that your custom dataset images and annotations are properly normalized and match the preprocessing expectations for YOLOv8.
> * **Batch Size and Learning Rate:** Sometimes, starting with a smaller learning rate or adjusting the batch size can help stabilize training and mitigate NaN issues.
> 
> As for MPS-specific suggestions:
> 
> * Double-check if all tensors are moved to the MPS device as expected. MPS can be a bit finicky about where tensors reside.
> * Look for any MPS-specific limitations or requirements in the Ultralytics documentation or PyTorch release notes.
> 
> Here’s a small tip for debugging, try isolating a small subset of your data (maybe just a few images) and see if training proceeds without NaNs. This can often point towards issues in the broader dataset.
> 
> Since you're using an M1 Pro, ensure your environment is up-to-date, and there are no conflicts between TensorFlow and PyTorch libraries, as your snippet seems to import both.
> 
> Remember, the community and Ultralytics team are always here to help. If the issue persists, feel free to share more details or reach out on the discussion boards! 💪

Hi @glenn-jocher ,
thank you for your answer. I double checked the annotations and those seem to be fine: 
![image](https://github.com/ultralytics/ultralytics/assets/38407239/ea70ca19-ed76-4000-8ac6-68f3c2472b14)

Classes from 0-3, x y w h .
As from the snippet, the parameters are in default state, i.e. batch = 16. I'll give it a try to lower the learning rate. I tried different values for batch size and image size with no effect. 

However, what worked for me: I did created a new enviroment, without tensorflow. Also, it installed python 3.12 instead of 3.9. I'm using a file for enviroment creation with python>=3.9. And with all that, it worked like a charm. There might me some issue with tensorflow and pytorch...

**Comment by mysakbm** [comment]
OK, found out that learning rate 0 will affect it. I reduced the lr0 to 0.001 and then the loss was quite huge but not a nan. However, after a while it get to nan nonetheless. It seems like a gradient exploding. Probably issue with labeling?

**Comment by glenn-jocher** [comment]
Hey @mysakbm 👋,

Great detective work on the learning rate! Yes, reducing `lr0` to a smaller value can indeed help avoid NaNs initially, pointing towards potential gradient issues.

Considering you still encounter NaNs after some time, it might be related to gradient instability which can sometimes arise from labeling issues, but could also be attributed to model or data specific characteristics. 

If you haven't already, you could try:
- Applying gradient clipping to avoid exploding gradients. This can be set in the training command with the `--gradient-accumulation` parameter.
- Double-checking the class labels in your dataset for correctness and consistency.
- Inspecting a few training examples manually to catch any obvious anomalies.

If your dataset and labels look correct, experimenting with further tuning the model's hyperparameters could also provide a path forward.

Keep pushing and feel free to reach out with any updates or further questions! 🚀

**Comment by mysakbm** [comment]
It was not just the learning rate. It seems to me like an alchemy class. 
I reduced the dataset to 33 images, checked them and bounding boxes seems ok. The images are quite large but I suppose that the parameter imgsz in train function takes care of it.
So, this is my whole code: 
```
import os
from ultralytics import YOLO

dataset_path = '../datasets/signatures_dataset/SignverOD/'
model = YOLO('yolov8n.pt')  # load a pretrained model (recommended for training)

# Train the model
results = model.train(
    data=dataset_path + '/signverOD.yaml',
    imgsz=96,
    lr0=0.001,  # initial learning rate
    epochs=2,
    val=False,  # disable validation
    batch = 4,
    device='mps'
)
```
What I tried to change is the 
- imgsz parameter from 640 -> 320 -> 170 -> 96. 
- epochs from 5 -> 2 -> 1
- val from True to False
- batch from 16 -> 8 -> 4 ->2 -> 1

And the problem is either that the loss is either NAN from the beginning or it goes for less than 5 epochs and then fails on "Invalid buffer size: 40.00 GB". If I reduce the batchsize to 1 or 2 then the error is invalid buffer size 80.00 GB.   


I'm completely lost. I turned off the validation after training but the function still calls the validator for some reason. And I suspect the validator to cause the issue. I'll dig into the validator function.
![image](https://github.com/ultralytics/ultralytics/assets/38407239/5d1b1215-adf8-4e14-9239-5ce07d440c13)

**Comment by mysakbm** [comment]
OK. Found it. The issue is definitely in the self.validation process. 
I deleted all images from the folder except for 14. Then the training process went without issues even with default lr, default batch size, or slightly bigger images (from 96 to 160) but lower the batch size. 
The issue might be that in yaml file i set up the validation folder the same as training folder, however, I thought that turning the validation off will be enough.

**Comment by glenn-jocher** [comment]
Hey @mysakbm, great job narrowing down the issue! 🕵️‍♂️ It sounds like you've pinpointed the validation process as the culprit. Yep, setting the validation data the same as your training data can sometimes lead to unexpected behaviors, especially if validation isn't explicitly turned off in every part of the process.

Even when you disable validation during training (`val=False`), ensure that the validation path in your YAML doesn't point to the same data as your training. It's generally a good practice to separate these datasets or, if you're aiming to train without validation, consider removing or commenting out the validation dataset path in your YAML file to avoid any confusion.

Here's a quick tip:
```yaml
# In your dataset.yaml, you can comment out the val section if not used
train: ../datasets/signatures_dataset/SignverOD/train/images
# val: ../datasets/signatures_dataset/SignverOD/val/images  # Comment this out if not validating
```

Keep up the good work, and don't hesitate to reach out if you have more questions! 🚀

**Comment by andreivicol** [comment]
> > @mysakbm hey there! 👋 It seems like you've encountered the infamous NaN issue when training on MPS with a custom dataset. Since training on the COCO dataset works fine, but your custom dataset leads to NaN values, this typically hints at issues such as label formatting or data preprocessing discrepancies specific to your custom dataset.
> > Here are a couple of things you could check:
> > 
> > * **Label Format:** Make sure your labels are correctly formatted according to YOLOv8 requirements. Label issues are a common cause of NaNs during training.
> > * **Normalization:** Verify that your custom dataset images and annotations are properly normalized and match the preprocessing expectations for YOLOv8.
> > * **Batch Size and Learning Rate:** Sometimes, starting with a smaller learning rate or adjusting the batch size can help stabilize training and mitigate NaN issues.
> > 
> > As for MPS-specific suggestions:
> > 
> > * Double-check if all tensors are moved to the MPS device as expected. MPS can be a bit finicky about where tensors reside.
> > * Look for any MPS-specific limitations or requirements in the Ultralytics documentation or PyTorch release notes.
> > 
> > Here’s a small tip for debugging, try isolating a small subset of your data (maybe just a few images) and see if training proceeds without NaNs. This can often point towards issues in the broader dataset.
> > Since you're using an M1 Pro, ensure your environment is up-to-date, and there are no conflicts between TensorFlow and PyTorch libraries, as your snippet seems to import both.
> > Remember, the community and Ultralytics team are always here to help. If the issue persists, feel free to share more details or reach out on the discussion boards! 💪
> 
> Hi @glenn-jocher , thank you for your answer. I double checked the annotations and those seem to be fine: ![image](https://private-user-images.githubusercontent.com/38407239/315744196-ea70ca19-ed76-4000-8ac6-68f3c2472b14.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NDkxMTcsIm5iZiI6MTcxMTQ0ODgxNywicGF0aCI6Ii8zODQwNzIzOS8zMTU3NDQxOTYtZWE3MGNhMTktZWQ3Ni00MDAwLThhYzYtNjhmM2MyNDcyYjE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDEwMjY1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljYWJlMWYxMTBlMWIxYTJiYTFhMTAzODYxMjA4NTE4OTc2OWI5ZTVlYWE2NWIwZjA0Y2FmZmYwNzA0ZGVkMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lTHQAfCAwh7K0-0Dzxsv3xCQxALphNzt056qBYuGFEs)
> 
> Classes from 0-3, x y w h . As from the snippet, the parameters are in default state, i.e. batch = 16. I'll give it a try to lower the learning rate. I tried different values for batch size and image size with no effect.
> 
> I created a new enviroment with ultralytics and pytorch from scratch. Still a nan. Could you please suggest a way to double check: if all tensors are moved to the MPS device as expected. Also, i don't understand what do you mean by this: Look for any MPS-specific limitations or requirements in the Ultralytics documentation or PyTorch release notes.

Hello, @mysakbm 

Thank you for sharing interest in my issue and delving deeper into the issue. I suspect I might have this labelling problem, would you mind sharing your script for labelling images and forming the dataset? I have, so far, been unable to resolve this.

Thank you!

**Comment by mysakbm** [comment]
Hi @andreivicol ,

I could definitely share my script, however, you have to take into account difference source we are getting the dataset. 
First, if you are asking me for labeling script then I have to disappoint you, because I have none. I used labelImg package to draw all the bounding boxes myself. And for some public datasets like Tabacco or SignverOD, they were already annotated and required no additional work from my side. However, those annotations could be in various formats: coco, yolo, pascal, etc. 
It's a good practice to check what labels format do you have. Just draw the bounding box.
```
image_idx = 2

image_path = '../datasets/signatures_dataset/SignverOD/' + 'images/'
label_path = '../datasets/signatures_dataset/SignverOD/' + 'labels/'

image_list = os.listdir(image_path)

with open(label_path + image_list[image_idx].split(".")[0] + ".txt", "r") as file:
    L = file.readlines()

# draw one bounding box just for sanity check
x, y, w, h = [float(x) for x in L[0].strip().split()[1:]]
x = x - w / 2
y = y - h / 2

image = cv2.imread(image_path + image_list[image_idx])
image_w = image.shape[1]
image_h = image.shape[0]

x = int(x * image_w)
y = int(y * image_h)
w = int(w * image_w)
h = int(h * image_h)

cv2.rectangle(image, (x, y), (x + w, y + h), (0, 255, 0), 2)

cv2.imshow('image', image)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

I expected that my bounding box is already in yolo format, i.e. 
- x and y is the center of the bounding box. 
- w and h is the width and heigh of the image 
- all those 4 values are relative to image size (i.e. between 0 and 1)

So, I substracted half of the width and heigh from the x and y to get lower left corner of the bounding box and multiplied values x,y,w,h with actual image size. Then, I plotted rectangle on the image. 

If the bounding box correspond to the label you expected then you have no problem with bounding boxes.

Third, I think that the problem cannot be inherently in the labeling. Because, what can happen with bad value? The loss will just be higher or lower. So, I think that the problem must be in the image size, batch size. If you use defaut values for the rest of the train function it will automatically determine the learning rate, even if you set up your own. So, I suggest to you to lower the image size to 96, batch size to 2-4, and try it. 

If you followed the convo above you saw that the issue is also a number of images for validation. On my MacBook Pro M1 with 16gb of ram, too many images in validation folder caused and buffer overflow. 

* so, check the bounding boxes
* elaborate on those image size parameters and try to use only handful of them - like 10 in the folder.
